### ISSUE TITLE
crash in optimizer

### ISSUE THREAD

**Issue by llvmbot** [llvm/llvm-project#24052]
|  |  |
| --- | --- |
| Bugzilla Link | [23678](https://llvm.org/bz23678) |
| Version | 3.6 |
| OS | MacOS X |
| Attachments | [compiler command](https://user-images.githubusercontent.com/60944935/143751500-849587b9-b898-4255-8171-1796e4619736.gz), [preprocessed cpp file](https://user-images.githubusercontent.com/60944935/143751501-776dc487-e05b-423e-926f-30bd022a97fe.gz) |
| Reporter | LLVM Bugzilla Contributor |

## Extended Description 
Just found a crash in clang that seems to be in the optimizer. The crash reproduces at -O2 and -O3, but not at -O1. It also reproduces if the opt tool is run on a .bc file emitted by the compiler without optimizations.

Here's the trace:

0  clang                    0x00000001058e8bf9 void std::__1::seed_seq::generate<unsigned int*>(unsigned int*, unsigned int*) + 10777
1  clang                    0x00000001058e935b void std::__1::seed_seq::generate<unsigned int*>(unsigned int*, unsigned int*) + 12667
2  libsystem_platform.dylib 0x00007fff869dff1a _sigtramp + 26
3  libsystem_platform.dylib 0x0000000106271a00 _sigtramp + 2139691776
4  clang                    0x0000000104ad6d8d llvm::SmallVectorTemplateBase<std::__1::unique_ptr<llvm::GlobalVariable, std::__1::default_delete<llvm::GlobalVariable> >, false>::grow(unsigned long) + 12909
5  clang                    0x00000001055c98b8 std::__1::__tree<std::__1::__value_type<llvm::Function const*, llvm::CallGraphNode*>, std::__1::__map_value_compare<llvm::Function const*, std::__1::__value_type<llvm::Function const*, llvm::CallGraphNode*>, std::__1::less<llvm::Function const*>, true>, std::__1::allocator<std::__1::__value_type<llvm::Function const*, llvm::CallGraphNode*> > >::destroy(std::__1::__tree_node<std::__1::__value_type<llvm::Function const*, llvm::CallGraphNode*>, void*>*) + 2728
6  clang                    0x00000001058798e8 void llvm::DeleteContainerSeconds<llvm::DenseMap<llvm::Type*, llvm::UndefValue*, llvm::DenseMapInfo<llvm::Type*>, llvm::detail::DenseMapPair<llvm::Type*, llvm::UndefValue*> > >(llvm::DenseMap<llvm::Type*, llvm::UndefValue*, llvm::DenseMapInfo<llvm::Type*>, llvm::detail::DenseMapPair<llvm::Type*, llvm::UndefValue*> >&) + 32296
7  clang                    0x0000000103c8ceaa void std::__1::vector<clang::serialization::ModuleFile*, std::__1::allocator<clang::serialization::ModuleFile*> >::__push_back_slow_path<clang::serialization::ModuleFile* const&>(clang::serialization::ModuleFile* const&&&) + 9450
8  clang                    0x0000000103d8ebf2 llvm::SmallVectorTemplateBase<llvm::DenseMap<clang::BaseSubobject, unsigned long long, llvm::DenseMapInfo<clang::BaseSubobject>, llvm::detail::DenseMapPair<clang::BaseSubobject, unsigned long long> >, false>::grow(unsigned long) + 27570
9  clang                    0x0000000103e10dc4 void std::__1::vector<clang::CodeGen::LValue, std::__1::allocator<clang::CodeGen::LValue> >::__push_back_slow_path<clang::CodeGen::LValue const&>(clang::CodeGen::LValue const&&&) + 1236
10 clang                    0x0000000103ae247e std::__1::__tree<std::__1::__value_type<clang::edit::FileOffset, clang::edit::EditedSource::FileEdit>, std::__1::__map_value_compare<clang::edit::FileOffset, std::__1::__value_type<clang::edit::FileOffset, clang::edit::EditedSource::FileEdit>, std::__1::less<clang::edit::FileOffset>, true>, std::__1::allocator<std::__1::__value_type<clang::edit::FileOffset, clang::edit::EditedSource::FileEdit> > >::destroy(std::__1::__tree_node<std::__1::__value_type<clang::edit::FileOffset, clang::edit::EditedSource::FileEdit>, void*>*) + 5630
11 clang                    0x0000000103ab4833 void std::__1::vector<clang::FrontendInputFile, std::__1::allocator<clang::FrontendInputFile> >::__push_back_slow_path<clang::FrontendInputFile const&>(clang::FrontendInputFile const&&&) + 20851
12 clang                    0x0000000103a777f6 void std::__1::__tree_balance_after_insert<std::__1::__tree_node_base<void*>*>(std::__1::__tree_node_base<void*>*, std::__1::__tree_node_base<void*>*) + 4022
13 clang                    0x0000000103a6f029
14 clang                    0x0000000103a739d0 void std::__1::vector<llvm::SourceMgr::SrcBuffer, std::__1::allocator<llvm::SourceMgr::SrcBuffer> >::__push_back_slow_path<llvm::SourceMgr::SrcBuffer>(llvm::SourceMgr::SrcBuffer&&) + 2048
15 libdyld.dylib            0x00007fff8b0b05c9 start + 1
Stack dump:
0.	Program arguments: /Users/mlogan/toolchain/clang+llvm-3.6.0/bin/clang -cc1 -triple x86_64-apple-macosx10.10.0 -emit-llvm-bc -disable-free -disable-llvm-verifier -main-file-name native-bindings.cpp -mrelocation-model pic -pic-level 2 -mthread-model posix -mdisable-fp-elim -masm-verbose -munwind-tables -target-cpu core2 -target-linker-version 241.9 -dwarf-column-info -coverage-file /Users/mlogan/artillery/projects/atlas/game/CPP/build-x86_64-apple-darwin/native-bindings.bc -resource-dir /Users/mlogan/toolchain/clang+llvm-3.6.0/bin/../lib/clang/3.6.0 -I entity-lib/../third-party/glm -I entity-lib/../third-party/crc32 -I build-x86_64-apple-darwin -I ../../../../local/include -I ../../../../native-client/deps/SDL2/include -I . -I Lib -I entity-lib -I ThirdParty/box2d.js/Box2D_v2.2.1 -I ../../../../native-client/src -I ../../../../native-client/deps/io.js/deps/v8/include -stdlib=libc++ -O2 -std=c++11 -fdeprecated-macro -fdebug-compilation-dir /Users/mlogan/artillery/projects/atlas/game/CPP -ferror-limit 19 -fmessage-length 176 -stack-protector 1 -mstackrealign -fblocks -fobjc-runtime=macosx-10.10.0 -fencode-extended-block-signature -fcxx-exceptions -fexceptions -fmax-type-align=16 -fdiagnostics-show-option -fcolor-diagnostics -vectorize-loops -vectorize-slp -o build-x86_64-apple-darwin/native-bindings.bc -x c++ build-x86_64-apple-darwin/native-bindings.cpp
1.	<eof> parser at end of file
2.	Per-module optimization passes
3.	Running pass 'CallGraph Pass Manager' on module 'build-x86_64-apple-darwin/native-bindings.cpp'.
clang: error: unable to execute command: Floating point exception: 8
clang: error: clang frontend command failed due to signal (use -v to see invocation)
clang version 3.6.0 (tags/RELEASE_360/final)
Target: x86_64-apple-darwin14.1.0
Thread model: posix
clang: note: diagnostic msg: PLEASE submit a bug report to http://llvm.org/bugs/ and include the crash backtrace, preprocessed source, and associated run script.
clang: note: diagnostic msg:
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang: note: diagnostic msg: /var/folders/4j/95xx1xm16xn20mqgqv4jd_xc0000gn/T/native-bindings-2ac66c.cpp
clang: note: diagnostic msg: /var/folders/4j/95xx1xm16xn20mqgqv4jd_xc0000gn/T/native-bindings-2ac66c.sh
clang: note: diagnostic msg:

********************

**Comment by Endilll** [comment]
I'm hitting the following assertion, which seems related:
```
clang: /home/tonik/llvm-project/llvm/lib/Transforms/Utils/InlineFunction.cpp:1143:
bool llvm::InlineFunction(llvm::CallSite, llvm::InlineFunctionInfo&, bool):
Assertion `AllocaArraySize > 0 && "array size of AllocaInst is zero"' failed.

#0 0x17eee62 llvm::sys::PrintStackTrace(_IO_FILE*) (/media/hdd2tb/compiler-explorer/clang-assertions-3.6.0/bin/clang-3.6+0x17eee62)
#1 0x17ed049 SignalHandler(int) Signals.cpp:0:0
#2 0x7f718fe5a510 (/lib/x86_64-linux-gnu/libc.so.6+0x3c510)
#3 0x7f718fea80fc __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
#4 0x7f718fe5a472 raise ./signal/../sysdeps/posix/raise.c:27:6
#5 0x7f718fe444b2 abort ./stdlib/abort.c:81:7
#6 0x7f718fe443d5 _nl_load_domain ./intl/loadmsgcat.c:1177:9
#7 0x7f718fe533a2 (/lib/x86_64-linux-gnu/libc.so.6+0x353a2)
#8 0x1815f8f llvm::InlineFunction(llvm::CallSite, llvm::InlineFunctionInfo&, bool) (/media/hdd2tb/compiler-explorer/clang-assertions-3.6.0/bin/clang-3.6+0x1815f8f)
#9 0x1553de2 InlineCallIfPossible(llvm::CallSite, llvm::InlineFunctionInfo&, llvm::DenseMap<llvm::ArrayType*, std::vector<llvm::AllocaInst*, std::allocator<llvm::AllocaInst*>>, llvm::DenseMapInfo<llvm::ArrayType*>, llvm::detail::DenseMapPair<llvm::ArrayType*, std::vector<llvm::AllocaInst*, std::allocator<llvm::AllocaInst*>>>>&, int, bool, llvm::DataLayout const*) Inliner.cpp:0:0
#10 0x1556053 llvm::Inliner::runOnSCC(llvm::CallGraphSCC&) (/media/hdd2tb/compiler-explorer/clang-assertions-3.6.0/bin/clang-3.6+0x1556053)
#11 0x1ec2828 (anonymous namespace)::CGPassManager::runOnModule(llvm::Module&) CallGraphSCCPass.cpp:0:0
#12 0x14e0d48 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/media/hdd2tb/compiler-explorer/clang-assertions-3.6.0/bin/clang-3.6+0x14e0d48)
#13 0x1c1d959 clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, llvm::raw_ostream*) (/media/hdd2tb/compiler-explorer/clang-assertions-3.6.0/bin/clang-3.6+0x1c1d959)
#14 0x1c0bdeb clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) crtstuff.c:0:0
#15 0x207be45 clang::ParseAST(clang::Sema&, bool, bool) (/media/hdd2tb/compiler-explorer/clang-assertions-3.6.0/bin/clang-3.6+0x207be45)
#16 0x1c0caae clang::CodeGenAction::ExecuteAction() (/media/hdd2tb/compiler-explorer/clang-assertions-3.6.0/bin/clang-3.6+0x1c0caae)
#17 0x197d366 clang::FrontendAction::Execute() (/media/hdd2tb/compiler-explorer/clang-assertions-3.6.0/bin/clang-3.6+0x197d366)
#18 0x1958d11 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/media/hdd2tb/compiler-explorer/clang-assertions-3.6.0/bin/clang-3.6+0x1958d11)
#19 0x1a06223 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/media/hdd2tb/compiler-explorer/clang-assertions-3.6.0/bin/clang-3.6+0x1a06223)
#20 0x94f228 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/media/hdd2tb/compiler-explorer/clang-assertions-3.6.0/bin/clang-3.6+0x94f228)
#21 0x926c38 main (/media/hdd2tb/compiler-explorer/clang-assertions-3.6.0/bin/clang-3.6+0x926c38)
#22 0x7f718fe456ca __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:74:3
#23 0x7f718fe45785 call_init ./csu/../csu/libc-start.c:128:20
#24 0x7f718fe45785 __libc_start_main ./csu/../csu/libc-start.c:347:5
#25 0x94b8f9 _start (/media/hdd2tb/compiler-explorer/clang-assertions-3.6.0/bin/clang-3.6+0x94b8f9)
Stack dump:
0.      Program arguments: /opt/compiler-explorer/clang-assertions-3.6.0/bin/clang -cc1 -triple x86_64-apple-macosx10.10.0 -emit-llvm-bc -disable-free -disable-llvm-verifier -main-file-name native-bindings.cpp -mrelocation-model pic -pic-level 2 -mthread-model posix -mdisable-fp-elim -masm-verbose -munwind-tables -target-cpu core2 -target-linker-version 241.9 -dwarf-column-info -stdlib=libc++ -O2 -std=c++11 -fdeprecated-macro -ferror-limit 19 -fmessage-length 176 -stack-protector 1 -mstackrealign -fblocks -fobjc-runtime=macosx-10.10.0 -fencode-extended-block-signature -fcxx-exceptions -fexceptions -fmax-type-align=16 -fdiagnostics-show-option -fcolor-diagnostics -vectorize-loops -vectorize-slp -x c++ native-bindings-2ac66c.cpp 
1.      <eof> parser at end of file
2.      Per-module optimization passes
3.      Running pass 'CallGraph Pass Manager' on module 'native-bindings-2ac66c.cpp'.
Aborted
```

**Comment by Endilll** [comment]
Appears to be fixed in Clang 3.7: https://godbolt.org/z/j1P9ajqEr
Reduced by C-Reduce:
```cpp
void V8_Wrap_Function_Call(int *[]);
void Call(int argc) {
  int *hiddenArgv[argc];
  V8_Wrap_Function_Call(hiddenArgv);
}
void (*installBuiltins___trans_tmp_3)() = []() {
  Call(0);
};
```
